# Trabajo de Fin de Máster: Applicación Web sobre un modelo de Question Answering en español

Este notebook recreará desde cero cómo realizar un **Fine-Tuning** de un modelo BERT para realizar un sistema de **Question Answering**.

A partir del código creado en [repositorio](https://github.com/jhackdue/tfm_app), importaremos las librerías necesarias y empezaremos a pasar por fases de configuración, preprocesado, elección de parámetros, creación del modelo, entrenamiento y validación, predicción.


Pero exactamente, ¿Cómo haremos que se entrene el modelo a partir de Bert?. Primero, miramos la estructura del modelo ajustado ya para este propósito de Question Answering:

<img src="bert.png" width=400 height=400 style="margin-left:auto; margin-right:auto"/>

En la memoria del TFM encontramos más detalle sobre esto, pero en resumidas cuentas:
- Configuraremos los textos de preguntas y contextos en una sola secuencia tokenizada
- Crearemos una red neuronal de dos neuronas que nos dirán dónde empiezan y acaba una respuesta dentro de los tokens del contexto.

La salida de Bert de cada token (Ti) multiplicado por el peso de la primera neurona (S) nos da la probabilidad de que ese token sea el inicio de la respuesta. Pasa lo mismo si multiplicamos por el peso de la segunda neurona (E), obtenemos la probabilidad de que otro token (Tj) sea el final de la respuesta. La puntuación de un intervalo candidato sería $S . Ti + E . Tj$, el modelo que entrenamos intenta maximizar esta puntuación.

---
## Índice General

* [Capítulo 0: Configuración Inicial](#cap0)

    * [Importación de librerías](#cap0.1)
    * [Configuración del proyecto](#cap0.2)
    * [visualización de datos](#cap0.3)

* [Capítulo 1: Preprocesado de datos](#cap1)

    * [Configurando el tokenizador](#cap1.1)
    * [Modificando la configuración](#cap1.2)
    * [Preprocesamiento](#cap1.3)
    
* [Capítulo 2: Construcción del modelo](#cap2)

    * [Hiperparámetros](#cap2.1)
    * [Callbacks](#cap2.2)
    * [Formato Keras](#cap2.3)
    * [Modelo](#cap2.4)
    
* [Capítulo 3: Entrenamiento y validación](#cap3)

    * [Monitorización en Tensorboard](#cap3.1)
    * [Entrenamiento](#cap3.2)
    
* [Capítulo 4: Predicción](#cap4)

    * [Carga de modelo](#cap4.1)
    * [Dato a evaluar](#cap4.2)
    * [Modelo entrenado](#cap4.3)
    * [Modelo remoto](#cap4.4)
    * [Pipeline](#cap4.5)

---
<a class="anchor" id="cap0"></a>
## Capítulo 0: Configuración Inicial

En esta parte del notebook empezaremos importando las librerías necesarias, posteriormente nos aseguraremos que las carpetas y ficheros creados etén en sus respectivos lugares y visualizaremos algunos ejemplos del dataset

<a class="anchor" id="cap0.1"></a>
### Importación de librerias 

- Importamos librerías como numpy y json para el posterior tratado de los datos
- Añadimos al entorno del notebook rutas fuera del directorio padre de este notebook
- Importamos las librerías necesarias para realizar la lectura, preprocesado, entrenamiento y predicción

In [1]:
import numpy as np
import json
import os
import sys
ABS_DIR = os.path.join(os.getcwd(), "..")
sys.path.append(ABS_DIR)


import utils.read_and_write as rw
import utils.preprocesado as pp
import train.train_utils as tu
import predict.predict_utils as pu

<a class="anchor" id="cap0.2"></a>
### Configuración del proyecto

* Descargamos los ficheros de entrenamiento y evalución de SQuAD en español desde [aquí](https://github.com/ccasimiro88/TranslateAlignRetrieve)
* Descargamos los ficheros de vocab y config con los que se entrenó BETO (Spanish BERT) desde [aquí](https://github.com/dccuchile/beto)
* Con las funciones de rw, comprobamos que existen los ficheros descargados en los directorios indicados

In [2]:
CONFIG_DIR = os.path.join(ABS_DIR, "config\\")

DATA_DIR = os.path.join(ABS_DIR, "data\\")
LOG_DIR = os.path.join(DATA_DIR, "logs\\")
MODELS_DIR = os.path.join(DATA_DIR, "models\\")
DATASETS_DIR = os.path.join(DATA_DIR, "datasets\\")
CHECKPOINTS_CALLBACK_DIR = os.path.join(DATA_DIR, "checkpoints\\")
CACHE_DIR = os.path.join(DATA_DIR, "cache\\")

logger = rw.crear_logger("tfm-app.log")

train_path = rw.comprobar_fichero_existe(os.path.join(DATASETS_DIR, "train-v2.0-es.json"), logger)
eval_path = rw.comprobar_fichero_existe(os.path.join(DATASETS_DIR, "dev-v2.0-es.json"), logger)

vocab_BERT_path = rw.comprobar_fichero_existe(os.path.join(MODELS_DIR, "vocab.txt"), logger)
config_BERT_path = rw.comprobar_fichero_existe(os.path.join(MODELS_DIR, "config.json"), logger)

config_file = rw.cargar_config()

2021-11-09 16:15:05,605 tfm-app.log INFO: El fichero log tfm-app.log ha sido creado
2021-11-09 16:15:05,624 tfm-app.log INFO: El fichero train-v2.0-es.json existe en D:\tfm_app\src\project\notebooks\..\data\datasets
2021-11-09 16:15:05,625 tfm-app.log INFO: El fichero dev-v2.0-es.json existe en D:\tfm_app\src\project\notebooks\..\data\datasets
2021-11-09 16:15:05,626 tfm-app.log INFO: El fichero vocab.txt existe en D:\tfm_app\src\project\notebooks\..\data\models
2021-11-09 16:15:05,628 tfm-app.log INFO: El fichero config.json existe en D:\tfm_app\src\project\notebooks\..\data\models


<a class="anchor" id="cap0.3"></a>
### Visualización de datos 

Podemos comprobar, a partir de los ficheros json, cómo son los datos tanto de entrenamiento como de validación. Para ello primero los pasaremos a un Pandas Dataframe

In [3]:
train_df = rw.json_to_dataframe(train_path)
print(train_df.shape)
eval_df = rw.json_to_dataframe(eval_path)
print(eval_df.shape)

(86818, 6)
(9905, 6)


* Ejemplo de datos de entrenamiento

In [4]:
id_row = 5000
print(f"Título:\n {train_df.iloc[id_row, 1]}")
print("---"*30)
print(f"Párrafo:\n {train_df.iloc[id_row, 2]}")
print("---"*30)
print(f"Pregunta:\n {train_df.iloc[id_row, 3]}")
print("---"*30)
print(f"Posicion de respuesta:\n {train_df.iloc[id_row, 4]}")
print("---"*30)
print(f"Respuesta:\n {train_df.iloc[id_row, 5]}")

Título:
 Budismo
------------------------------------------------------------------------------------------
Párrafo:
 La segunda verdad es que el origen de la dukkha puede ser conocido. En el contexto de las cuatro verdades nobles, el origen de dukkha se explica comúnmente como deseo (Pali: Tanha) condicionado por la ignorancia (Pali: Avijja). En un nivel más profundo, la causa raíz del dukkha se identifica como ignorancia (Pali: Avijja) de la verdadera naturaleza de las cosas. La tercera verdad noble es que el cese completo de la dukkha es posible, y la cuarta verdad noble identifica un camino a este cese. [Nota 7]
------------------------------------------------------------------------------------------
Pregunta:
 El origen del dukkha se explica como deseo condicionado por qué?
------------------------------------------------------------------------------------------
Posicion de respuesta:
 201
------------------------------------------------------------------------------------------

* Ejemplo de datos de validación

In [5]:
id_row = 5000
print(f"Título:\n {eval_df.iloc[id_row, 1]}")
print("---"*30)
print(f"Párrafo:\n {eval_df.iloc[id_row, 2]}")
print("---"*30)
print(f"Pregunta:\n {eval_df.iloc[id_row, 3]}")
print("---"*30)
print(f"Posicion de respuesta:\n {eval_df.iloc[id_row, 4]}")
print("---"*30)
print(f"Respuesta:\n {eval_df.iloc[id_row, 5]}")

Título:
 Escuela privada
------------------------------------------------------------------------------------------
Párrafo:
 Las escuelas religiosas afiliadas y denominacionales forman una subcategoría de escuelas privadas. Algunas de estas escuelas enseñan educación religiosa, junto con los sujetos académicos habituales para impresionar las creencias y tradiciones de su fe particular en los estudiantes que asisten. Otros usan la denominación como una etiqueta más general para describir lo que los fundadores basaron en su creencia, manteniendo una fina distinción entre académicos y religión. Incluyen escuelas parroquiales, un término que a menudo se utiliza para denotar escuelas católicas romanas. Otros grupos religiosos representados en el sector de la educación privada K-12 incluyen protestantes, judíos, musulmanes y cristianos ortodoxos.
------------------------------------------------------------------------------------------
Pregunta:
 Junto con musulmanes, judíos y cristianos pr

---
<a class="anchor" id="cap1"></a>
## Capítulo 1: Preprocesado de datos

En este capítulo obtendremos los datos que entrarán directamente en el modelo

<a class="anchor" id="cap1.1"></a>
### Configurando el tokenizador

El tokenizador que se va a utilizar se creará a partir del vocabulario de tokens que tenemos en el directorio del modelo

In [6]:
tokenizador = pp.obtener_tokenizador(vocab=vocab_BERT_path, lowercase=False)

Por ejemplo, podemos tokenizar la siguiente frase y ver qué tokens asigna nuestro tokenizador:

In [7]:
tokens = tokenizador.encode("Esto es una prueba para ver cómo se tokeniza")
print(f'ids: {tokens.ids}')
print(f'tokens: {tokens.tokens}')

ids: [4, 2098, 1058, 1108, 3889, 1110, 1292, 2078, 1062, 3, 5]
tokens: ['[CLS]', 'Esto', 'es', 'una', 'prueba', 'para', 'ver', 'cómo', 'se', '[UNK]', '[SEP]']


Vemos que cuenta con casi todos los tokens, excepto por la palabra "tokenizar" que no lo tenía en el vocabulario y pasa a ser el token [UNK]. También observamos que añade tanto al principio como al final, dos tokens [CLS] y [SEP]

<a class="anchor" id="cap1.2"></a>
### Modificando la configuración

Tenemos la configuración de la aplicación para entrenar que cuenta con:

In [8]:
config_file

{'logs': {'level': 'DEBUG',
  'format': '%(asctime)s %(name)s %(levelname)s: %(message)s'},
 'train': {'preprocess': {'max_seq_len': 384,
   'lr': 0.05,
   'batch_size': 64,
   'nb_epoch': 20}}}

Lo que haremos será añadir la parte del preprocesado a la configuración de Bert que tenemos en el directorio del modelo

In [9]:
with open(config_BERT_path, 'r') as json_file:
    config_BERT = json.load(json_file)
config_BERT.update(config_file["train"]["preprocess"])

<a class="anchor" id="cap1.3"></a>
### Preprocesamiento

En esta parte del notebook, lo que vamos a hacer es, dado los datos de entrenamiento y validación en un DataFrame de pandas, convertirlos en un formato de tensorflow. En particular obtendremos:
* x = [id_tokens, token_type, attention_mask]
* y = [id_start_token, id_end_token]
* dataset = tf.Dataset.from_tensor_slices(({id_tokens: tf.cast, token_type: tf.cast, attention_mask: tf.cast}, {id_start_token: tf.cast, id_end_token: tf.cast}))
* squad_Objects = [Objetos de tipo SquadExample]
* Errores = [Id del del DF que no ha sido seleccionado para entrenamiento/validación]

Se puede encontrar más información sobre el formato tf.Dataset [aquí](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_tensor_slices). Uno de los motivos de que se generen errores al transformar a un objeto de tipo SquadExample, es porque el contexto y la pregunta suman más de 384 tokens, que es la secuencia máxima permitida.

* Entrenamiento: Convertimos los datos de tipo DF a tf.Dataset y también creamos los SquadExample de los datos de entrenamiento

In [9]:
x_train, y_train, train_dataset, train_squad_objects, train_errors = pp.transformar_datos_squad(train_df, tokenizador, config_BERT, logger=logger, name_data="entrenamiento")

2021-10-14 23:27:57,657 tfm-app.log INFO: Transformado el conjunto de entrenamiento al formato SquadExample
2021-10-14 23:28:03,531 tfm-app.log INFO: Se ha conseguido obtener los inputs y los targets para el conjunto de entrenamiento. Se han creado  85629 puntos


* Validación: Convertimos los datos de tipo DF a tf.Dataset y también creamos los SquadExample de los datos de validación

In [10]:
x_eval, y_eval, eval_dataset, eval_squad_objects, eval_errors = pp.transformar_datos_squad(eval_df, tokenizador, config_BERT, logger=logger, name_data="validación")

2021-10-14 23:28:10,812 tfm-app.log INFO: Transformado el conjunto de validación al formato SquadExample
2021-10-14 23:28:11,388 tfm-app.log INFO: Se ha conseguido obtener los inputs y los targets para el conjunto de validación. Se han creado  9649 puntos


---
<a class="anchor" id="cap2"></a>
## Capítulo 2: Construcción del modelo

Llegamos a la parte donde iremos construyendo lo necesario para el entrenamiento de nuestro modelo y también la validación

<a class="anchor" id="cap2.1"></a>
### Hiperparámetros

Tenemos por tanto la configuración del modelo BERT y de cómo vamos a entrenar el modelo ajustado para el Question Answering.

In [9]:
config_BERT

{'attention_probs_dropout_prob': 0.1,
 'hidden_act': 'gelu',
 'hidden_dropout_prob': 0.1,
 'hidden_size': 768,
 'initializer_range': 0.02,
 'intermediate_size': 3072,
 'max_position_embeddings': 512,
 'num_attention_heads': 12,
 'num_hidden_layers': 12,
 'type_vocab_size': 2,
 'vocab_size': 31002,
 'max_seq_len': 384,
 'lr': 0.05,
 'batch_size': 64,
 'nb_epoch': 20}

In [10]:
logger.info(f"Modelo entrenado con los siguientes hiperparámetros: {config_BERT}")

2021-10-31 19:31:16,381 tfm-app.log INFO: Modelo entrenado con los siguientes hiperparámetros: {'attention_probs_dropout_prob': 0.1, 'hidden_act': 'gelu', 'hidden_dropout_prob': 0.1, 'hidden_size': 768, 'initializer_range': 0.02, 'intermediate_size': 3072, 'max_position_embeddings': 512, 'num_attention_heads': 12, 'num_hidden_layers': 12, 'type_vocab_size': 2, 'vocab_size': 31002, 'max_seq_len': 384, 'lr': 0.05, 'batch_size': 64, 'nb_epoch': 20}


<a class="anchor" id="cap2.2"></a>
### Callbacks

Dentro del entrenamiento, queremos generar untos de control para poder ir almacenando en cada epoch el mejor modelo, centrarnos en minimizar la función de coste lo más rápido posible, poder visualizar las métricas que nos están saliendo para ver si estamos mejorando o no en la pérdida.

In [12]:
callbacks = tu.generar_callbacks(x_eval, y_eval, eval_squad_objects, logger=logger)

<a class="anchor" id="cap2.3"></a>
### Formato Keras

Convertimos todo el tf.Dataset en un conjunto de batches, que viene configurado en config_BERT

In [13]:
train_dataset_keras = pp.input_formato_keras(train_dataset, config_BERT)

<a class="anchor" id="cap2.4"></a>
### Modelo

Se inicializa el modelo con toda la construcción inicial. Por debajo lleva librerías de Transformers de HuggingFace

In [10]:
modelo = tu.obtener_modelo(logger=logger, config=config_BERT)

Some layers from the model checkpoint at dccuchile/bert-base-spanish-wwm-cased were not used when initializing TFBertForQuestionAnswering: ['mlm___cls']
- This IS expected if you are initializing TFBertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some layers of TFBertForQuestionAnswering were not initialized from the model checkpoint at dccuchile/bert-base-spanish-wwm-cased and are newly initialized: ['qa_outputs']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
2021-10-17 14:14:52,587 tfm-app.log INFO: Se ha realizado la carga del modelo

Se puede apreciar cómo se ha congelado la capa base, donde tenemos el modelo Bert. Y tendremos la salida en el formato de índices de tokens donde está el inicio y fin de repuesta.

In [19]:
modelo.summary()

Model: "tf_bert_for_question_answering"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
bert (TFBertMainLayer)       multiple                  109850880 
_________________________________________________________________
qa_outputs (Dense)           multiple                  1538      
Total params: 109,852,418
Trainable params: 1,538
Non-trainable params: 109,850,880
_________________________________________________________________


---
<a class="anchor" id="cap3"></a>
## Capítulo 3: Entrenamiento y validación

Iniciaremos el método fit de entrenamiento del modelo, pero antes, iniciamos el tensorboard

<a class="anchor" id="cap3.1"></a>
### Monitorización en Tensorboard

Hemos generado un callback específico del tensorboard para poder ir mirando las métricas asociadas al entrenamiento. En este caso, la función de coste

In [10]:
# Cargamos el tensorboard
%load_ext tensorboard
# %reload_ext tensorboard

In [12]:
%tensorboard --logdir '.\..\\data\\logs\\tensorboard\\'

Reusing TensorBoard on port 6006 (pid 1916), started 0:00:56 ago. (Use '!kill 1916' to kill it.)

<a class="anchor" id="cap3.2"></a>
### Entrenamiento

Le pasamos el modelo, el dataset en formato de batches, la configuración del modelo, los callbacks generados y el fichero de logs para poder ir controlando el entrenamiento.

In [21]:
modeloHistory = tu.entrenar_modelo(modelo, train_dataset_keras, config_BERT, callbacks, logger=logger)

2021-10-14 23:30:23,287 tfm-app.log INFO: Se realizará el entrenamiento con 20 epochs


Epoch 1/20
   1/1338 [..............................] - ETA: 0s - loss: 11.8042 - output_1_loss: 5.9276 - output_2_loss: 5.8766WARNING:tensorflow:From d:\python_entornos\tf_gpu\lib\site-packages\tensorflow\python\ops\summary_ops_v2.py:1277: stop (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.stop` instead.
1338/1338 [==============================] - ETA: 0s - loss: 7.9015 - output_1_loss: 3.9903 - output_2_loss: 3.9112
Epoch 00001: loss improved from inf to 7.90154, saving model to ..\data\checkpoints\model.01-7.90.h5


2021-10-15 01:44:29,703 tfm-app.log INFO: 
epoch=1, exact match score=0.00



epoch=1, exact match score=0.00
1338/1338 [==============================] - 8030s 6s/step - loss: 7.9015 - output_1_loss: 3.9903 - output_2_loss: 3.9112
Epoch 2/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8848 - output_1_loss: 3.9877 - output_2_loss: 3.8971
Epoch 00002: loss improved from 7.90154 to 7.88476, saving model to ..\data\checkpoints\model.02-7.88.h5


2021-10-15 03:58:28,133 tfm-app.log INFO: 
epoch=2, exact match score=0.00



epoch=2, exact match score=0.00
1338/1338 [==============================] - 8033s 6s/step - loss: 7.8848 - output_1_loss: 3.9877 - output_2_loss: 3.8971
Epoch 3/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8930 - output_1_loss: 3.9771 - output_2_loss: 3.9160
Epoch 00003: loss did not improve from 7.88476


2021-10-15 06:12:16,195 tfm-app.log INFO: 
epoch=3, exact match score=0.00



epoch=3, exact match score=0.00
1338/1338 [==============================] - 8022s 6s/step - loss: 7.8930 - output_1_loss: 3.9771 - output_2_loss: 3.9160
Epoch 4/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8792 - output_1_loss: 3.9715 - output_2_loss: 3.9076
Epoch 00004: loss improved from 7.88476 to 7.87917, saving model to ..\data\checkpoints\model.04-7.88.h5


2021-10-15 08:26:06,606 tfm-app.log INFO: 
epoch=4, exact match score=0.00



epoch=4, exact match score=0.00
1338/1338 [==============================] - 8025s 6s/step - loss: 7.8792 - output_1_loss: 3.9715 - output_2_loss: 3.9076
Epoch 5/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8852 - output_1_loss: 3.9794 - output_2_loss: 3.9058
Epoch 00005: loss did not improve from 7.87917


2021-10-15 10:39:57,931 tfm-app.log INFO: 
epoch=5, exact match score=0.00



epoch=5, exact match score=0.00
1338/1338 [==============================] - 8026s 6s/step - loss: 7.8852 - output_1_loss: 3.9794 - output_2_loss: 3.9058
Epoch 6/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8759 - output_1_loss: 3.9697 - output_2_loss: 3.9062
Epoch 00006: loss improved from 7.87917 to 7.87586, saving model to ..\data\checkpoints\model.06-7.88.h5


2021-10-15 12:53:59,360 tfm-app.log INFO: 
epoch=6, exact match score=0.00



epoch=6, exact match score=0.00
1338/1338 [==============================] - 8036s 6s/step - loss: 7.8759 - output_1_loss: 3.9697 - output_2_loss: 3.9062
Epoch 7/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8915 - output_1_loss: 3.9788 - output_2_loss: 3.9128
Epoch 00007: loss did not improve from 7.87586


2021-10-15 15:08:04,875 tfm-app.log INFO: 
epoch=7, exact match score=0.00



epoch=7, exact match score=0.00
1338/1338 [==============================] - 8040s 6s/step - loss: 7.8915 - output_1_loss: 3.9788 - output_2_loss: 3.9128
Epoch 8/20
1338/1338 [==============================] - ETA: 0s - loss: 7.9044 - output_1_loss: 3.9816 - output_2_loss: 3.9228
Epoch 00008: loss did not improve from 7.87586


2021-10-15 17:22:07,800 tfm-app.log INFO: 
epoch=8, exact match score=0.00



epoch=8, exact match score=0.00
1338/1338 [==============================] - 8037s 6s/step - loss: 7.9044 - output_1_loss: 3.9816 - output_2_loss: 3.9228
Epoch 9/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8640 - output_1_loss: 3.9720 - output_2_loss: 3.8920
Epoch 00009: loss improved from 7.87586 to 7.86398, saving model to ..\data\checkpoints\model.09-7.86.h5


2021-10-15 19:36:12,966 tfm-app.log INFO: 
epoch=9, exact match score=0.00



epoch=9, exact match score=0.00
1338/1338 [==============================] - 8040s 6s/step - loss: 7.8640 - output_1_loss: 3.9720 - output_2_loss: 3.8920
Epoch 10/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8849 - output_1_loss: 3.9742 - output_2_loss: 3.9107
Epoch 00010: loss did not improve from 7.86398


2021-10-15 21:50:15,184 tfm-app.log INFO: 
epoch=10, exact match score=0.00



epoch=10, exact match score=0.00
1338/1338 [==============================] - 8037s 6s/step - loss: 7.8849 - output_1_loss: 3.9742 - output_2_loss: 3.9107
Epoch 11/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8931 - output_1_loss: 3.9822 - output_2_loss: 3.9109
Epoch 00011: loss did not improve from 7.86398


2021-10-16 00:04:26,239 tfm-app.log INFO: 
epoch=11, exact match score=0.00



epoch=11, exact match score=0.00
1338/1338 [==============================] - 8045s 6s/step - loss: 7.8931 - output_1_loss: 3.9822 - output_2_loss: 3.9109
Epoch 12/20
1338/1338 [==============================] - ETA: 0s - loss: 7.8745 - output_1_loss: 3.9726 - output_2_loss: 3.9019
Epoch 00012: ReduceLROnPlateau reducing learning rate to 0.005000000074505806.

Epoch 00012: loss did not improve from 7.86398


2021-10-16 02:18:17,994 tfm-app.log INFO: 
epoch=12, exact match score=0.00



epoch=12, exact match score=0.00
1338/1338 [==============================] - 8026s 6s/step - loss: 7.8745 - output_1_loss: 3.9726 - output_2_loss: 3.9019
Epoch 13/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0420 - output_1_loss: 3.5600 - output_2_loss: 3.4820
Epoch 00013: loss improved from 7.86398 to 7.04197, saving model to ..\data\checkpoints\model.13-7.04.h5


2021-10-16 04:32:13,791 tfm-app.log INFO: 
epoch=13, exact match score=0.01



epoch=13, exact match score=0.01
1338/1338 [==============================] - 8030s 6s/step - loss: 7.0420 - output_1_loss: 3.5600 - output_2_loss: 3.4820
Epoch 14/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0091 - output_1_loss: 3.5444 - output_2_loss: 3.4647
Epoch 00014: loss improved from 7.04197 to 7.00909, saving model to ..\data\checkpoints\model.14-7.01.h5


2021-10-16 06:46:02,752 tfm-app.log INFO: 
epoch=14, exact match score=0.01



epoch=14, exact match score=0.01
1338/1338 [==============================] - 8023s 6s/step - loss: 7.0091 - output_1_loss: 3.5444 - output_2_loss: 3.4647
Epoch 15/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0101 - output_1_loss: 3.5423 - output_2_loss: 3.4678
Epoch 00015: loss did not improve from 7.00909


2021-10-16 08:59:52,589 tfm-app.log INFO: 
epoch=15, exact match score=0.01



epoch=15, exact match score=0.01
1338/1338 [==============================] - 8024s 6s/step - loss: 7.0101 - output_1_loss: 3.5423 - output_2_loss: 3.4678
Epoch 16/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0116 - output_1_loss: 3.5453 - output_2_loss: 3.4663
Epoch 00016: loss did not improve from 7.00909


2021-10-16 11:13:43,037 tfm-app.log INFO: 
epoch=16, exact match score=0.01



epoch=16, exact match score=0.01
1338/1338 [==============================] - 8025s 6s/step - loss: 7.0116 - output_1_loss: 3.5453 - output_2_loss: 3.4663
Epoch 17/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0064 - output_1_loss: 3.5423 - output_2_loss: 3.4642
Epoch 00017: loss improved from 7.00909 to 7.00644, saving model to ..\data\checkpoints\model.17-7.01.h5


2021-10-16 13:27:36,401 tfm-app.log INFO: 
epoch=17, exact match score=0.01



epoch=17, exact match score=0.01
1338/1338 [==============================] - 8028s 6s/step - loss: 7.0064 - output_1_loss: 3.5423 - output_2_loss: 3.4642
Epoch 18/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0067 - output_1_loss: 3.5415 - output_2_loss: 3.4651
Epoch 00018: loss did not improve from 7.00644


2021-10-16 15:41:33,783 tfm-app.log INFO: 
epoch=18, exact match score=0.01



epoch=18, exact match score=0.01
1338/1338 [==============================] - 8032s 6s/step - loss: 7.0067 - output_1_loss: 3.5415 - output_2_loss: 3.4651
Epoch 19/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0027 - output_1_loss: 3.5422 - output_2_loss: 3.4605
Epoch 00019: loss improved from 7.00644 to 7.00272, saving model to ..\data\checkpoints\model.19-7.00.h5


2021-10-16 17:55:25,923 tfm-app.log INFO: 
epoch=19, exact match score=0.01



epoch=19, exact match score=0.01
1338/1338 [==============================] - 8027s 6s/step - loss: 7.0027 - output_1_loss: 3.5422 - output_2_loss: 3.4605
Epoch 20/20
1338/1338 [==============================] - ETA: 0s - loss: 7.0103 - output_1_loss: 3.5438 - output_2_loss: 3.4664
Epoch 00020: loss did not improve from 7.00272


2021-10-16 20:09:19,993 tfm-app.log INFO: 
epoch=20, exact match score=0.00



epoch=20, exact match score=0.00
1338/1338 [==============================] - 8028s 6s/step - loss: 7.0103 - output_1_loss: 3.5438 - output_2_loss: 3.4664


2021-10-16 20:09:20,172 tfm-app.log INFO: Entrenamiento finalizado!
2021-10-16 20:09:20,248 tfm-app.log INFO: Guardando pesos...
2021-10-16 20:09:21,484 tfm-app.log INFO: Se han guardado los pesos en: ..\data\models\qa_model_squad_v2_esp\qa_model_squad_v2_esp.h5
2021-10-16 20:09:21,485 tfm-app.log INFO: Guardando Modelo Json...
2021-10-16 20:09:21,487 tfm-app.log ERROR: 


---
<a class="anchor" id="cap4"></a>
## Capítulo 4: Predicción

Una vez tenemos nuestro modelo entrenado, procedemos a inferir nuestras respuestas a partir de una pregunta y un contexto

<a class="anchor" id="cap4.1"></a>
### Carga de modelo

Cargamos los modelos para realizar una inferencia sobre ellos. Tenemos nuestro modelo que hemos entrenado, pero también tenemos en local el [modelo distill Bert](https://huggingface.co/mrm8488/distill-bert-base-spanish-wwm-cased-finetuned-spa-squad2-es) que se utiliza en la aplicación, puesto que se ha entrenado en mejores capacidades computacionales

* Nuestro Modelo

In [13]:
modelo = tu.cargar_modelo(logger=logger, config=config_BERT)

Some layers from the model checkpoint at dccuchile/bert-base-spanish-wwm-cased were not used when initializing TFBertForQuestionAnswering: ['mlm___cls']
- This IS expected if you are initializing TFBertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some layers of TFBertForQuestionAnswering were not initialized from the model checkpoint at dccuchile/bert-base-spanish-wwm-cased and are newly initialized: ['qa_outputs']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
2021-11-09 17:01:51,545 tfm-app.log INFO: Se ha cargado el modelo qa_model_sq

In [14]:
modelo.summary()

Model: "tf_bert_for_question_answering"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
bert (TFBertMainLayer)       multiple                  109850880 
_________________________________________________________________
qa_outputs (Dense)           multiple                  1538      
Total params: 109,852,418
Trainable params: 109,852,418
Non-trainable params: 0
_________________________________________________________________


* Modelo Remoto

In [18]:
modelo_remoto = tu.cargar_modelo(model_name="remote_model", logger=logger)

Some layers from the model checkpoint at dccuchile/bert-base-spanish-wwm-cased were not used when initializing TFBertForQuestionAnswering: ['mlm___cls']
- This IS expected if you are initializing TFBertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some layers of TFBertForQuestionAnswering were not initialized from the model checkpoint at dccuchile/bert-base-spanish-wwm-cased and are newly initialized: ['qa_outputs']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
2021-11-09 17:03:42,280 tfm-app.log INFO: Se ha cargado el modelo remote_mode

In [19]:
modelo_remoto.summary()

Model: "tf_bert_for_question_answering_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
bert (TFBertMainLayer)       multiple                  109850880 
_________________________________________________________________
qa_outputs (Dense)           multiple                  1538      
Total params: 109,852,418
Trainable params: 109,852,418
Non-trainable params: 0
_________________________________________________________________


<a class="anchor" id="cap4.2"></a>
### Dato a evaluar

Elegimos de entre los datos de evaluación, un contexto y una pregunta y comparamos con la respuesta de verdad

In [20]:
eval_df.reset_index(drop=True, inplace=True)
print(len(eval_df))

9905


In [21]:
ejemplo_id = 772
context = eval_df.loc[ejemplo_id, "Context"]
print("Contexto:")
print(context, "\n")

question = eval_df.loc[ejemplo_id, "Question"]
print("Pregunta:")
print(question, "\n")

respuesta = eval_df.loc[ejemplo_id, "Text"]
print("Respuesta:")
print(respuesta)

Contexto:
El servicio comenzó el 1 de septiembre de 1993 basado en la idea del entonces director ejecutivo, Sam Chisholm y Rupert Murdoch, de convertir la estrategia de negocio de la compañía a un concepto completamente basado en la fiebre. El nuevo paquete incluye cuatro canales que antes estaban disponibles de forma gratuita, transmitiendo en los satélites de Astra, así como la introducción de nuevos canales. El servicio continuó hasta el cierre del servicio analógico de BSkyB el 27 de septiembre de 2001, debido al lanzamiento y expansión de la plataforma Sky Digital. Algunos de los canales transmitieron ya sea en el cifrado claro o suave (por lo que se requería un decodificador VideoCrypt para decodificar, sin una tarjeta de suscripción) antes de su adición al paquete Sky Multichannels. A los dos meses del lanzamiento, BSkyB ganó 400,000 nuevos suscriptores, con la mayoría tomando al menos un canal premium también, lo que ayudó a BSkyB llegar a 3,5 millones de hogares a mediados de 

<a class="anchor" id="cap4.3"></a>
### Modelo entrenado

Primero, predecimos con nuestro modelo entrenado

In [22]:
respuesta_predicha = pu.predict(question, context, modelo=modelo, tokenizer=tokenizador)

In [23]:
respuesta_predicha

{'start': 3,
 'end': 96,
 'answer': 'el 1 de septiembre de 1993 basado en la idea del entonces director ejecutivo, Sam Chisholm y Rupert Murdoch, de convertir la estrategia de negocio de la compañía a un concepto completamente basado en la fiebre. El nuevo paquete incluye cuatro canales que antes estaban disponibles de forma gratuita, transmitiendo en los satélites de Astra, así como la introducción de nuevos canales. El servicio continuó hasta el cierre del servicio analógico de BSkyB el 27 de septiembre de 2001, debido al lanzamiento y expansión de la plataforma Sky Digital. Algunos de los canales transmitieron'}

In [27]:
from IPython.core.display import HTML

marked_text = str(context.replace(respuesta_predicha['answer'], f"<mark>{respuesta_predicha['answer']}</mark>"))
HTML(f"<p>{question.upper()}</p>\n" + f"""<blockquote> {marked_text} </blockquote>""")

<a class="anchor" id="cap4.4"></a>
### Modelo Remoto

Seguimos con el modelo en remoto, pero que lo tenemos descargado en local

In [28]:
respuesta_predicha_remoto = pu.predict(question, context, modelo=modelo_remoto, tokenizer=tokenizador)

In [29]:
respuesta_predicha_remoto

{'start': 67, 'end': 68, 'answer': 'continuó hasta'}

In [30]:
from IPython.core.display import HTML

marked_text = str(context.replace(respuesta_predicha_remoto['answer'], f"<mark>{respuesta_predicha_remoto['answer']}</mark>"))
HTML(f"<p>{question.upper()}</p>\n" + f"""<blockquote> {marked_text} </blockquote>""")

<a class="anchor" id="cap4.5"></a>
### Pipeline

Aunque, por si hubiera algún error de formato al descargarlo en local, podemos inferir directamente haciendo peticiones HTTP al repositorio del [modelo de huggingFace](https://huggingface.co/mrm8488/distill-bert-base-spanish-wwm-cased-finetuned-spa-squad2-es)

In [31]:
respuesta_predicha_pipeline = pu.predict(question, context, use_pipeline=True)

In [32]:
respuesta_predicha_pipeline

{'score': 1.104228203985258e-06,
 'start': 341,
 'end': 359,
 'answer': 'satélites de Astra'}

In [33]:
from IPython.core.display import HTML

marked_text = str(context.replace(respuesta_predicha_pipeline['answer'], f"<mark>{respuesta_predicha_pipeline['answer']}</mark>"))
HTML(f"<p>{question.upper()}</p>\n" + f"""<blockquote> {marked_text} </blockquote>""")

---